code/game/sound.dm

```diff
@@ -22,6 +26,65 @@
 		0, // Flags (1 = Auto Direct, 2 = Auto Room, 4 = Auto RoomHF)
 	)
 	environment = SOUND_ENVIRONMENT_NONE //Default to none so sounds without overrides dont get reverb
+	var/atom/attached
+	var/list/listeners = list()
+	var/maxdist = 0
+
+/sound/proc/attach( var/atom/towhat )
+	attached = towhat
+	channel = GLOB.channelReserved++
+	if(GLOB.channelReserved>GLOB.channelMax)
+		GLOB.channelReserved=256
+	if(!attached.attachedSounds)
+		attached.attachedSounds = list()
+	attached.attachedSounds.Add(src)
+	GLOB.allAttachedSounds.Add(src)
+
+	recalculate()
+
+/sound/proc/stop()
+	if(channel)
+		if(!GLOB.mutecache)
+			GLOB.mutecache = sound(null)
+		GLOB.mutecache.channel = channel
+		world << src
+		if(attached)
+			attached.attachedSounds.Remove( src )
+			GLOB.allAttachedSounds.Remove( src )
+			attached = null
+			listeners = list()
+
+/sound/proc/recalcClient(var/client/c)
+	if(!GLOB.mutecache)
+		GLOB.mutecache = sound(null)
+	GLOB.mutecache.channel = channel
+	status |= SOUND_UPDATE
+	var/turf/CT = get_turf(c.mob)
+	var/turf/AT = get_turf(attached)
+	var/dist = get_dist(CT, AT)
+	if( !listeners.Find(c) )
+		//world << "Their dist: [dist] vs [maxdist]"
+		if( dist < maxdist )
+			status &= ~SOUND_UPDATE
+			pan = clamp((AT.x - CT.x)/maxdist*100,-100,100)
+			volume = 100-(dist/maxdist*100)
+			SEND_SOUND(c, src)
+			status |= SOUND_UPDATE
+			listeners.Add(c)
+
+	else if( dist > maxdist )
+		listeners.Remove(c)
+		SEND_SOUND(c, GLOB.mutecache)
+	else
+		pan = (AT.x - CT.x)/maxdist*100
+		volume = 100-(dist/maxdist*100)
+		SEND_SOUND(c, src)
+
+/sound/proc/recalculate()
+	for(var/client/c in GLOB.clients)
```

Comment:
better var name, please

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names